primitives/session-validator-management/query/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+use main_chain_follower_api::CandidateDataSource;
+use plutus::ToDatum;
+use sidechain_block_search::{predicates::AnyBlockInEpoch, FindSidechainBlock, SidechainInfo};
+use sidechain_domain::{MainchainPublicKey, McEpochNumber, ScEpochNumber};
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::{HeaderBackend, Info};
+use sp_core::bytes::to_hex;
+use sp_runtime::traits::NumberFor;
+use sp_runtime::traits::{Block as BlockT, Zero};
+use sp_session_validator_management::SessionValidatorManagementApi;
+use sp_sidechain::{GetSidechainParams, GetSidechainStatus};
+use std::sync::Arc;
+use types::*;
+
+pub type QueryResult<T> = Result<T, String>;
+
+#[async_trait]
+pub trait SessionValidatorManagementQueryApi {
+	/// Returns the committee for given sidechain epoch. The order of the list represents the order of slot allocation.
+	fn get_epoch_committee(&self, epoch_number: u64) -> QueryResult<GetCommitteeResponse>;
+
+	///
+	/// returns: Last active and valid registration followed by all newer invalid registrations for mc_epoch_number and mc_public_key.
+	/// Regardless of `mc_epoch_number` value, it always uses validation api from the latest sidechain block.
+	///
+	async fn get_registrations(
+		&self,
+		mc_epoch_number: McEpochNumber,
+		mc_public_key: MainchainPublicKey,
+	) -> QueryResult<Vec<CandidateRegistrationEntry>>;
+
+	/// Regardless of `epoch_number` value, all the candidates data validation is done based on the validation api from the latest sidechain block.
+	async fn get_ariadne_parameters(
+		&self,
+		epoch_number: McEpochNumber,
+	) -> QueryResult<AriadneParameters>;
+}
+
+#[derive(new)]
+pub struct SessionValidatorManagementQuery<
+	C,
+	Block,
+	SessionKeys: parity_scale_codec::Decode,
+	CrossChainPublic,
+	SidechainParams: parity_scale_codec::Decode + ToDatum + Clone + Send + Sync + 'static,
+> {
+	client: Arc<C>,
+	candidate_data_source: Arc<dyn CandidateDataSource + Send + Sync>,
+	_marker: std::marker::PhantomData<(Block, SessionKeys, CrossChainPublic, SidechainParams)>,
+}
+
+#[async_trait]
+impl<
+		C,
+		Block,
+		SessionKeys: parity_scale_codec::Decode + Send + Sync + 'static,
+		SidechainParams: parity_scale_codec::Decode + ToDatum + Clone + Send + Sync + 'static,
+		CrossChainPublic: parity_scale_codec::Decode
+			+ parity_scale_codec::Encode
+			+ AsRef<[u8]>
+			+ Send
+			+ Sync
+			+ 'static,
+	> SessionValidatorManagementQueryApi
+	for SessionValidatorManagementQuery<C, Block, SessionKeys, CrossChainPublic, SidechainParams>
+where
+	Block: BlockT,
+	NumberFor<Block>: From<u32> + Into<u32>,
+	C: Send + Sync + 'static,
+	C: ProvideRuntimeApi<Block>,
+	C: HeaderBackend<Block>,
+	C::Api: sp_api::Core<Block>,
+	C::Api: GetSidechainStatus<Block>,
+	C::Api: SessionValidatorManagementApi<
+		Block,
+		SessionKeys,
+		CrossChainPublic,
+		AuthoritySelectionInputs,
+		ScEpochNumber,
+	>,
+	C::Api: GetSidechainParams<Block, SidechainParams>,
+	C::Api: CandidateValidationApi<Block>,
+{
```

Comment:
Methods in this implementation are like methods that were in RPC, only errors are mapped to strings instead of jsonrpsee ErrorObjects.